src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -38,28 +38,142 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
+                    // Allow prerelease packages only if the target is prerelease
+                    var includePrelease = userAction.PackageIdentity.Version.IsPrerelease || userAction.Action == NuGetProjectActionType.Uninstall;
+                    var includeUnlisted = userAction.Action == NuGetProjectActionType.Uninstall;
+
+                    var resolutionContext = new ResolutionContext(uiService.DependencyBehavior, includePrelease, includeUnlisted, VersionConstraints.None);
+
+                    return GetActionsAsync(
+                        uiService,
+                        projects,
+                        userAction,
+                        uiService.RemoveDependencies,
+                        uiService.ForceRemove,
+                        resolutionContext,
+                        projectContext: uiService.ProgressWindow,
+                        token: token);
+                },
+                (actions) =>
+                {
+                    return ExecuteActionsAsync(actions, uiService.ProgressWindow, userAction, token);
+                },
+                windowOwner,
+                token);
+        }
+
+        /// <summary>
+        /// Perform the multi-package update action.
+        /// </summary>       
+        /// <remarks>This needs to be called from a background thread. It may hang on the UI thread.</remarks>
+        public async Task PerformUpdateAsync(
+            INuGetUI uiService,
+            List<PackageIdentity> packagesToUpdate,
+            DependencyObject windowOwner,
+            CancellationToken token)
+        {
+            await PerformActionImplAsync(
+                uiService,
+                () =>
+                {
+                    return ResolveActionsForUpdate(
+                        uiService,
+                        packagesToUpdate,
+                        token);
+                },
+                async (actions) =>
+                {
+                    foreach (var projectActions in actions.GroupBy(action => action.Item1))
+                    {
+                        await _packageManager.ExecuteNuGetProjectActionsAsync(
+                           projectActions.Key,
+                           projectActions.Select(action => action.Item2),
+                           uiService.ProgressWindow,
+                           token);
+                    }
+                },
+                windowOwner,
+                token);
+        }
+
+        /// <summary>
+        /// Calculates the list of actions needed to perform packages updates.
+        /// </summary>
+        /// <param name="uiService">ui service.</param>
+        /// <param name="packagesToUpdate">The list of packages to update.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns>The list of actions.</returns>
+        private async Task<IReadOnlyList<Tuple<NuGetProject, NuGetProjectAction>>> ResolveActionsForUpdate(
```

Comment:
Please do not use Tuples. It is against the coding guidelines. Consider using a private class